### PR TITLE
Telegram adapter

### PR DESCRIPTION
**Pull request by jasonharrison**
WIP Telegram adapter

To do:

- [x] Fix highlights/mentions (see `InviteHandler`)
- [x] Prefix (`.` for Discord, `/` for Telegram)

### CODE REVIEW

File: src/handlers/InviteHandler.ts [hunk 1]
```diff
@@ -86,10 +86,10 @@ export default class InviteHandler extends MessageHandler {
     await save(this.state);
     const invitees = players
       .filter(player => player.id !== creator.id)
-      .map(invitee => invitee.username)
+      .map(invitee => '[' + invitee.username + '](tg://user?id=' + invitee.id.value + ')')
       .join(', ');
     return this.simpleReply(
-      `${invitees}, do you want to play ${gameCode} with <@${creator.id.value}> ? Use ".accept" to accept, or ".reject" to reject invite.`,
+      `${invitees}, do you want to play ${gameCode} with @${creator.username} ? Use ".accept" to accept, or ".reject" to reject invite.`,
```
Comment (flamecoals) [inline src/handlers/InviteHandler.ts]: Oh, this is wrong. We should be using the simpleReply with mentions instead. The way we have right now wont work on both platforms

File: src/telegram/TelegramMessagingEnvironment.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+  constructor(client: TelegramBot) {
+    super();
+    this.client = client;
+  }
+
+  async isAdmin(user: User, channel: Channel) {
+    // TODO stub
+    return true;
+    // assert(user.id.namespace === Namespace.DISCORD);
```
Comment (flamecoals) [inline src/telegram/TelegramMessagingEnvironment.ts]: can you delete this commented out code?

File: src/util.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { Player, GameChannel } from './state';
 import { Id } from './id';
 import { Canvas } from 'canvas';
 
-export const PREFIX = '.';
+export const PREFIX = '/'; // FIXME temporary
```
Comment (flamecoals) [inline src/util.ts]: Yeah we need to fix this... Needs to be a variable in the MessagingEnvironment I think, and we need to also replace all replies right now hard coded with "."

File: src/telegram/adaptor.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  }
+  return mentions;
+}
+
+export function translateTelegramChannelType(channelType: string): ChannelType {
+  switch (channelType) {
+    case 'private':
+      return ChannelType.PVT;
+    case 'group':
+      return ChannelType.PUBLIC_GROUP;
```
Comment (flamecoals) [inline src/telegram/adaptor.ts]: It should be private group.... I dont think there are public groups in telegram?

File: src/telegram/adaptor.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+function sendMessageToUser(client: TelegramBot, user: TelegramBot.User, message: ReplyMessage) {
+  client.sendMessage(user.id, encodeMentions(message));
+}
+
+export function sendReplyToTelegram(client: TelegramBot, msg: TelegramBot.Message, reply: Reply) {
+  for (const message of reply.messages) {
+    if (message.type == ChannelType.PVT) {
+      sendMessageToUser(client, msg.from, message);
+    } else if (message.type == ChannelType.PUBLIC_GROUP) {
```
Comment (flamecoals) [inline src/telegram/adaptor.ts]: It should be private group.... I dont think there are public groups in telegram?

Comment (jasonharrison) [reply src/telegram/adaptor.ts]: I think there are public groups you can join with a link.  I'm honestly not sure how to differentiate between them.

**Pull request merged**